h1b/src/hil/flash/driver.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// The interface between the flash driver and the (real or fake) flash module.
+pub trait HwInterface {
+    /// Returns true if an operation is running, false otherwise.
+    fn is_programming(&self) -> bool;
+
+    /// Read a single word from the flash (non-blocking). offset is in units of
+    /// words and is relative to the start of flash.
+    fn read(&self, offset: usize) -> u32;
+
+    /// Reads the flash error code.
+    fn read_error(&self) -> u16;
+
+    /// Set flash transaction parameters (word offset and size). The word offset
+    /// is relative to the start of flash and the size is in units of words.
+    fn set_transaction(&self, offset: usize, size: usize);
+
+    /// Fill the flash controller's write buffer. data must have a length no
+    /// larger than 32.
+    fn set_write_data(&self, data: &[u32]);
+
+    /// Kick off a smart program execution.
+    fn trigger(&self, opcode: u32);
+}
+
+/// A client of the Flash driver -- receives callbacks when flash operations
+/// complete.
+pub trait Client {
+    fn erase_done(&self, kernel::ReturnCode);
```

Comment:
It's problematic to have callbacks that don't easily correlate with calls. I.e., if you have erase_done and write_done then there should be erase() and write() calls. Are these two traits at different levels of abstraction (write and erase are implemented on top of HwInterface)?

Reply:
I separated HwInterface into its own module to make the distinction between HwInterface and the driver more clear.

I'll implement erase() in the future, for now only write() is supported.